crates/bevy_render/src/erased_render_asset.rs

```diff
@@ -0,0 +1,430 @@
+        self.0.get_mut(&id.into())
+    }
+
+    pub fn insert(&mut self, id: impl Into<UntypedAssetId>, value: ERA) -> Option<ERA> {
+        self.0.insert(id.into(), value)
+    }
+
+    pub fn remove(&mut self, id: impl Into<UntypedAssetId>) -> Option<ERA> {
+        self.0.remove(&id.into())
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (UntypedAssetId, &ERA)> {
+        self.0.iter().map(|(k, v)| (*k, v))
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (UntypedAssetId, &mut ERA)> {
+        self.0.iter_mut().map(|(k, v)| (*k, v))
+    }
+}
+
+#[derive(Resource)]
+struct CachedExtractErasedRenderAssetSystemState<A: ErasedRenderAsset> {
+    state: SystemState<(
+        EventReader<'static, 'static, AssetEvent<A::SourceAsset>>,
+        ResMut<'static, Assets<A::SourceAsset>>,
+    )>,
+}
+
+impl<A: ErasedRenderAsset> FromWorld for CachedExtractErasedRenderAssetSystemState<A> {
+    fn from_world(world: &mut bevy_ecs::world::World) -> Self {
+        Self {
+            state: SystemState::new(world),
+        }
+    }
+}
+
+/// This system extracts all created or modified assets of the corresponding [`ErasedRenderAsset::SourceAsset`] type
+/// into the "render world".
+pub(crate) fn extract_erased_render_asset<A: ErasedRenderAsset>(
+    mut commands: Commands,
+    mut main_world: ResMut<MainWorld>,
+) {
+    main_world.resource_scope(
+        |world, mut cached_state: Mut<CachedExtractErasedRenderAssetSystemState<A>>| {
+            let (mut events, mut assets) = cached_state.state.get_mut(world);
+
+            let mut needs_extracting = <HashSet<_>>::default();
+            let mut removed = <HashSet<_>>::default();
+            let mut modified = <HashSet<_>>::default();
+
+            for event in events.read() {
+                #[expect(
+                    clippy::match_same_arms,
+                    reason = "LoadedWithDependencies is marked as a TODO, so it's likely this will no longer lint soon."
+                )]
+                match event {
+                    AssetEvent::Added { id } => {
+                        needs_extracting.insert(*id);
+                    }
+                    AssetEvent::Modified { id } => {
+                        needs_extracting.insert(*id);
+                        modified.insert(*id);
+                    }
+                    AssetEvent::Removed { .. } => {
+                        // We don't care that the asset was removed from Assets<T> in the main world.
+                        // An asset is only removed from ErasedRenderAssets<T> when its last handle is dropped (AssetEvent::Unused).
+                    }
+                    AssetEvent::Unused { id } => {
+                        needs_extracting.remove(id);
+                        modified.remove(id);
+                        removed.insert(*id);
+                    }
+                    AssetEvent::LoadedWithDependencies { .. } => {
+                        // TODO: handle this
```

Comment:
Is this for a follow-up?

Reply:
This file is basically a copy/paste of the existing render asset file, so this came from there.